src/diffusers/models/unets/unet_i2vgen_xl.py

```diff
@@ -0,0 +1,816 @@
+# Copyright 2023 Alibaba DAMO-VILAB and The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import torch
+import torch.nn as nn
+import torch.utils.checkpoint
+
+from ...configuration_utils import ConfigMixin, register_to_config
+from ...loaders import UNet2DConditionLoadersMixin
+from ...utils import deprecate, logging
+from ..activations import get_activation
+from ..attention import Attention, FeedForward
+from ..attention_processor import (
+    ADDED_KV_ATTENTION_PROCESSORS,
+    CROSS_ATTENTION_PROCESSORS,
+    AttentionProcessor,
+    AttnAddedKVProcessor,
+    AttnProcessor,
+)
+from ..embeddings import TimestepEmbedding, Timesteps
+from ..modeling_utils import ModelMixin
+from ..transformer_temporal import TransformerTemporalModel
+from .unet_3d_blocks import (
+    CrossAttnDownBlock3D,
+    CrossAttnUpBlock3D,
+    DownBlock3D,
+    UNetMidBlock3DCrossAttn,
+    UpBlock3D,
+    get_down_block,
+    get_up_block,
+)
+from .unet_3d_condition import UNet3DConditionOutput
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+def _to_tensor(inputs, device):
+    if not torch.is_tensor(inputs):
+        # TODO: this requires sync between CPU and GPU. So try to pass `inputs` as tensors if you can
+        # This would be a good case for the `match` statement (Python 3.10+)
+        is_mps = device.type == "mps"
+        if isinstance(inputs, float):
+            dtype = torch.float32 if is_mps else torch.float64
+        else:
+            dtype = torch.int32 if is_mps else torch.int64
+        inputs = torch.tensor([inputs], dtype=dtype, device=device)
+    elif len(inputs.shape) == 0:
+        inputs = inputs[None].to(device)
+
+    return inputs
+
+
+def _collapse_frames_into_batch(sample: torch.Tensor) -> torch.Tensor:
+    batch_size, channels, num_frames, height, width = sample.shape
+    sample = sample.permute(0, 2, 1, 3, 4).reshape(batch_size * num_frames, channels, height, width)
+
+    return sample
```

Comment:
I think this function is only used twice and can be written more or less as a one-liner -> can we remove it and just write out the function instead?

Reply:
I think the main reason why we did it this way is because the shapes of the `sample` is different for the two instances this method is being used. 

So, we have to unpack the shape into `batch_size, channels, num_frames, height, width` each time. So, made sense to delegate this utility.

Reply:
This function is used twice and can be written as a one-liner:
```py
sample = sample.permute(0, 2, 1, 3, 4).reshape(sample.shape[0] * sample.shape[2], sample.shape[1], sample.shape[3], sample.shape[4])
```

=> better to not have this function